sql/table.h

```diff
@@ -1882,6 +1883,69 @@ typedef struct st_foreign_key_info
   LEX_CSTRING *referenced_key_name;
   List<LEX_CSTRING> foreign_fields;
   List<LEX_CSTRING> referenced_fields;
+private:
+  unsigned char *fields_nullable= nullptr;
+
+  /**
+    Get the number of fields exist in foreign key relationship
+  */
+  unsigned get_n_fields() const noexcept
+  {
+    unsigned n_fields= foreign_fields.elements;
+    if (n_fields == 0)
+      n_fields= referenced_fields.elements;
+    return n_fields;
+  }
+
+public:
+  /**
+    Assign nullable field for referenced and foreign fields
+    based on number of fields. This nullable fields
+    should be allocated by engine for passing the
+    foreign key information
+    @param thd thread to allocate the memory
+    @param num_fields number of fields
+  */
+  void assign_nullable(THD *thd, unsigned num_fields) noexcept
+  {
+    if (!fields_nullable)
+      fields_nullable=
+        (unsigned char *)thd_calloc(thd,
+                                    my_bits_in_bytes(2 * num_fields));
+  }
+
+  /**
+    Set nullable bit for the field in the given field
+    @param referenced set null bit for referenced column
+    @param field field number
+    @param n_fields number of fields
+  */
+  void set_nullable(bool referenced, unsigned field,
+                    unsigned n_fields) noexcept
```

Comment:
It would seem to be safer to remove the parameter `n_fields` and remember the `num_fields` parameter that was passed to `assign_nullable()`. This object has a short lifetime and is not part of time-critical code.

I see that every call `assign_nullable()` is followed by a conditional call to `set_nullable()`. It looks like we could use a single member function that is a combination of the two. The member current function `set_nullable()` is only missing the parameter `thd` that is needed when we need to set a value to `true` and the object has not been allocated previously.